rmf_fleet_adapter/src/read_only/FleetAdapterNode.cpp

```diff
@@ -234,6 +243,40 @@ void FleetAdapterNode::update_robot(
   const RobotState& state,
   const ScheduleEntries::iterator& it)
 {
+  if (it->second->route.has_value())
+  {
+    auto& participant = it->second->schedule->participant();
+    uint64_t route_id = participant.itinerary().size() - 1;
+    uint64_t last_checkpoint_reached = participant.reached()[route_id];
+
+    uint64_t checkpoint_id = 0;
+    for (const auto& wp : it->second->route.value().trajectory())
+    {
+      if (checkpoint_id <= last_checkpoint_reached)
+      {
+        checkpoint_id++;
+        continue;
+      }
+
+      Eigen::Vector2d current_location =
+        Eigen::Vector2d(state.location.x, state.location.y);
+      Eigen::Vector2d checkpoint_pose =
+        Eigen::Vector2d(wp.position()[0], wp.position()[1]);
+      Eigen::Vector2d diff = current_location - checkpoint_pose;
+      // TODO(@xiyuoh) Make this merge_waypoint_distance configurable
+      if (diff.norm() < 0.3)
```

Comment:
Being within some distance of the waypoint is not a robust criteria. There are two opposite issues to consider:
1. 30cm can be a considerable distance in tight spaces. If a robot gets stopped 30cm from its goal but we report it as reached, that could trigger a waiting robot to approach prematurely, thinking that this read-only robot is completely out of the way already.
2. If a robot manages to move through the 30cm radius before a state update takes place then we won't see that it reached the point until it reaches the next one, causing avoidable delays.

This also isn't quite the right location to identify where a robot has reached. At this point in the code we don't know if the robot is even following the same path as its last update, so we might report reaching a waypoint of an out-of-date path.

In the function `handle_delay` we examine whether the path of the new state update matches the path that was previously put into the traffic schedule. [This](https://github.com/open-rmf/rmf_ros2/blob/12cf7832d3e8464509128052134d9af43f46807f/rmf_fleet_adapter/src/read_only/FleetAdapterNode.cpp#L313-L314) is where we verify that the waypoints in the state message match the tail of the waypoints for this robot in the traffic schedule.

[This point in handle_delay](https://github.com/open-rmf/rmf_ros2/blob/12cf7832d3e8464509128052134d9af43f46807f/rmf_fleet_adapter/src/read_only/FleetAdapterNode.cpp#L440) is where we update the delay to the previously reported path, so that's where we should be reporting the most recently reached waypoint. Since we've already verified that the new and old paths match by that point in the code, we can actually do a simple difference between the number of waypoints in the route and the number of waypoints in the remaining path as reported by the state message. It would look like:

```
const auto route_size = it->second->route.trajectory().size();
const auto remaining_path_size = state.path.size();
if (route_size > remaining_path_size)
{
  participant.reached(
    participant.current_plan_id(),
    0,
    route_size - remaining_path_size - 1);
}
```

The `- 1` is because we should only report that we have "reached" points which are no longer included in the remaining path of the state message.

Reply:
These are extremely good points; using the difference in number of waypoints left in the path makes a lot more sense. Thank you for pointing me to this solution, I've added them in [379d93b](https://github.com/open-rmf/rmf_ros2/pull/387/commits/379d93b59ab6630213dd1a993e91bf4db883d92b) and tested in simulation.